frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,414 @@
+  AlertmanagerRoute,
+} from '../alert-manager-config';
+import { RoutingLabelEditor } from './routing-labels-editor';
+import * as PagerDutyForm from './pagerduty-receiver-form';
+import * as WebhookForm from './webhook-receiver-form';
+
+/**
+ * Converts routes of a specific Receiver:
+ * {
+ *   receiver: "MyReceiver",
+ *   match: {
+ *     severity: "warning",
+ *     cluster: "myCluster"
+ *   },
+ *   match_re: {
+ *    service: "$foobar"
+ *  }
+};
+ * ...to array of labels for Routing Labels Editor component
+ * [
+ *   {
+ *     "name": "severity",
+ *     "value": "warning",
+ *     "isRegex": false
+ *   },
+ *   {
+ *     "name": "cluster",
+ *     "value": "myCluster",
+ *     "isRegex": false
+ *   },
+ *   {
+ *     "name": "service",
+ *     "value": "$foobar",
+ *     "isRegex": true
+ *   }
+ * ]
+ */
+const convertReceiverRoutesToEditorLabels = (
+  receiver: AlertmanagerReceiver,
+  routes: AlertmanagerRoute[],
+): RouteEditorLabel[] => {
+  if (!receiver) {
+    return [];
+  }
+
+  const routesOfReceiver = _.find(
+    routes,
+    (aRoute: AlertmanagerRoute) => aRoute.receiver === receiver.name,
+  );
+  const matches = _.map(routesOfReceiver?.match || {}, (v, k) => {
+    return { name: k, value: v, isRegex: false };
+  });
+  const regexMatches = _.map(routesOfReceiver?.match_re || {}, (v, k) => {
+    return { name: k, value: v, isRegex: true };
+  });
+  return _.concat([], matches, regexMatches);
+};
+
+/**
+ * Returns new Route object
+ * Ex:
+ * {
+ *   receiver: myNewReceiver,
+ *   match: {
+ *     "severity": "warning",
+ *     "cluster": "myCluster"
+ *   }
+ *   match_re {
+ *     "service": "^(foo1|foo2|baz)$",
+ *   }
+ * }
+ */
+const createRoute = (
+  receiver: AlertmanagerReceiver,
+  routeLabels: RouteEditorLabel[],
+): AlertmanagerRoute => {
+  return _.reduce(
+    routeLabels,
+    (acc, label) => {
+      _.set(acc, [label.isRegex ? 'match_re' : 'match', label.name], label.value);
+      return acc;
+    },
+    { receiver: receiver.name },
+  );
+};
+
+/**
+ * Returns new Receiver object
+ * Ex:
+ * {
+ *   name: MyNewReceiver
+ *   pagerduty_configs: {
+ *     routing_key: <integration_key>
+ *   }
+ * }
+ */
+const createReceiver = (
+  formValues,
+  createReceiverConfig: Function,
+  receiverToEdit: AlertmanagerReceiver,
+): AlertmanagerReceiver => {
+  const receiverConfig = createReceiverConfig(
+    formValues,
+    receiverToEdit && receiverToEdit[formValues.receiverType]
+      ? receiverToEdit[formValues.receiverType][0] // pass in receiver config if editing existing receiver
+      : {},
+  );
+  return {
+    name: formValues.receiverName,
+    [formValues.receiverType]: [{ ...receiverConfig }],
+  };
+};
+
+const subformFactory = (receiverType: string) => {
+  switch (receiverType) {
+    case 'pagerduty_configs':
+      return PagerDutyForm;
+    case 'webhook_configs':
+      return WebhookForm;
+    default:
+      return null;
+  }
+};
+
+function formReducer(formValues, action) {
```

Comment:
```suggestion
const formReducer = (formValues, action) => {
```

Would be nice to have types here

Reply:
fixed